core/trino-main/src/main/java/io/trino/execution/scheduler/ScaledWriterScheduler.java

```diff
@@ -87,18 +95,33 @@ public ScheduleResult schedule()
         return new ScheduleResult(done.get(), writers, future, WRITER_SCALING, 0);
     }
 
-    private int getNewTaskCount()
+    @VisibleForTesting
+    int getNewTaskCount()
     {
         if (scheduledNodes.isEmpty()) {
             return 1;
         }
 
-        double fullTasks = sourceTasksProvider.get().stream()
-                .filter(task -> !task.getState().isDone())
+        List<TaskInfo> runningTasks = sourceTasksProvider.get().stream()
+                .filter(task -> !task.getTaskStatus().getState().isDone())
+                .collect(Collectors.toList());
+
+        // When there is a big data skewness, there could be a bottleneck due to the skewed workers even if most of the workers are not over-utilized
+        double weightedN = 0.0;
+        double weightedSum = 0.0;
+        for (TaskInfo task : runningTasks) {
+            double weight = (double) task.getStats().getOutputPositions() / SCALE_WEIGHT_BASE;
+            weightedN += weight;
+            weightedSum += task.getTaskStatus().getOutputBufferStatus().isOverutilized() ? weight : 0.0;
+        }
+
+        double fullTasks = max(weightedN > 0.0 ? weightedSum / weightedN : 0,
```

Comment:
IIUIC after output positions reaches `SCALE_WEIGHT_BASE` or 1m rows, we won't rely on the old way of computing buffer utilization. So, I'm thinking maybe it is better to remove the old way and simply rely on the weighted version. Additionally, if we remove the old then I guess there's no need for `SCALE_WEIGHT_BASE` right?

Another thing, maybe it will be much cleaner to simply keep the weight as `outputDataSize`. It makes it much easier to understand since if a task is skewed then its output data size outweighs all others, and it'll be consistent with the use of `physicalWrittenBytes` in the other check. WDYT?

Reply:
I would like to keep the old way to avoid any regressions(e.g. many smaller tasks' output buffers are over-utilized due to big skewed tasks whose output buffers are not yet over-utilized). I still can remove `SCALE_WEIGHT_BASE`.

Using `outputDataSize` in lieu of `outputPositions' makes sense to me.